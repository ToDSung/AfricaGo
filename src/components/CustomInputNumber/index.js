import React, { useCallback, useRef } from 'react';

import PropTypes from 'prop-types';

import {
  BACKSPACE,
  DOWN_ARROW,
  ESC,
  KEY_CODES,
  NUMBER_KEY_CODES,
  UP_ARROW,
} from '../../constants/keycode';
import { NumberValidator } from '../../utils/validator';
import Box from '../Box';

const preventDefault = e => {
  e.preventDefault();
};

const CustomInputNumber = ({
  min = 0,
  max = Number.MAX_SAFE_INTEGER,
  step = 1,
  name,
  value = 0,
  disabled = false,
  onChange = () => {},
  onBlur = () => {},
}) => {
  const numberBoxRef = useRef(null);

  const focusNumberBoxRef = () => {
    numberBoxRef.current.focus();
  };

  const blurNumberBoxRef = () => {
    numberBoxRef.current.blur();
  };

  const checkNewValueValid = useCallback(
    newValue => {
      return NumberValidator({ min, max })(newValue);
    },
    [min, max]
  );

  const handleChange = newValue => {
    checkNewValueValid(newValue) &&
      onChange({
        target: {
          name,
          value: newValue,
        },
      });
  };

  const handleMinus = () => {
    focusNumberBoxRef();
    handleChange(+value - step);
  };

  const handlePlus = () => {
    focusNumberBoxRef();
    handleChange(+value + step);
  };

  const handleKeyDown = event => {
    focusNumberBoxRef();

    const { keyCode } = event;

    if (!(keyCode in KEY_CODES)) {
      return;
    }

    if ([DOWN_ARROW].includes(KEY_CODES[keyCode])) {
      return handleMinus();
    }

    if ([UP_ARROW].includes(KEY_CODES[keyCode])) {
      return handlePlus();
    }

    if (keyCode in NUMBER_KEY_CODES) {
      handleChange(+(`${value}` + `${KEY_CODES[keyCode]}`));
    }

    if (KEY_CODES[keyCode] === ESC) {
      blurNumberBoxRef();
    }

    if (KEY_CODES[keyCode] === BACKSPACE) {
      handleChange(+`${value}`.slice(0, -1));
    }
  };

  const handleBlur = () => {
    onBlur({
      target: {
        name,
        value,
      },
    });
  };

  return (
    <div className='flex gap-2 p-2'>
      {disabled ? (
        <>
          <Box className='select-none border-gray-600 text-4xl text-gray-600'>
            -
          </Box>
          <Box className='relative border-gray-600 outline-none'>{value}</Box>
          <Box className='select-none border-gray-600 text-4xl text-gray-600'>
            +
          </Box>
        </>
      ) : (
        <>
          <input type='number' onBlur={onBlur}>
            {/* {value} */}
          </input>
          <Box
            onMouseDown={preventDefault} // to avoid focus this element
            onClick={handleMinus}
            className='cursor-pointer select-none border-sky-600  text-4xl  text-sky-600'
          >
            -
          </Box>
          <Box
            ref={numberBoxRef}
            onBlur={handleBlur}
            onKeyDown={handleKeyDown}
            className='relative border-sky-600 outline-none focus:border-2	'
            tabIndex={0}
          >
            {value}
          </Box>
          <Box
            onMouseDown={preventDefault}
            onClick={handlePlus}
            className='cursor-pointer select-none border-sky-600 text-5xl text-sky-600	'
          >
            +
          </Box>
        </>
      )}
    </div>
  );
};

CustomInputNumber.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  step: PropTypes.number,
  name: PropTypes.string.isRequired,
  value: PropTypes.number,
  disabled: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  onBlur: PropTypes.func,
};

export default CustomInputNumber;
