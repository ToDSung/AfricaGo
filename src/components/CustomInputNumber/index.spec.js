/**
 * @jest-environment jsdom
 */

import React from 'react';
import { act, create } from 'react-test-renderer';

import { fireEvent, render } from '@testing-library/react';

import CustomInputNumber from '.';

test('should CustomInputNumber render disabled', () => {
  const mockHandleChange = () => jest.fn();

  const component = create(
    <CustomInputNumber
      min={0}
      max={4}
      step={1}
      name='test'
      value={2}
      disabled={true}
      onChange={mockHandleChange(0)}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  act(() => {
    component.update(
      <CustomInputNumber
        min={0}
        max={4}
        step={1}
        name='test'
        value={3}
        disabled={true}
        onChange={mockHandleChange(0)}
      />
    );
  });

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('should CustomInputNumber render active', () => {
  const mockHandleChange = () => jest.fn();

  const component = create(
    <CustomInputNumber
      min={0}
      max={4}
      step={1}
      name='test'
      value={2}
      disabled={false}
      onChange={mockHandleChange(0)}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  act(() => {
    component.update(
      <CustomInputNumber
        min={0}
        max={4}
        step={1}
        name='test'
        value={3}
        disabled={false}
        onChange={mockHandleChange(0)}
      />
    );
  });

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('should CustomInputNumber Content disabled', () => {
  const mockHandleChange = jest.fn();

  const { queryByText } = render(
    <CustomInputNumber
      min={0}
      max={4}
      step={1}
      name='test'
      value={2}
      disabled={true}
      onChange={mockHandleChange}
    />
  );

  fireEvent.click(queryByText(/-/i));
  expect(mockHandleChange).toHaveBeenCalledTimes(0);
  fireEvent.click(queryByText(/\+/i));
  expect(mockHandleChange).toHaveBeenCalledTimes(0);
});

test('should CustomInputNumber Content active', () => {
  const mockHandleChange = jest.fn();

  const { queryByText } = render(
    <CustomInputNumber
      min={0}
      max={4}
      step={1}
      name='test'
      value={2}
      disabled={false}
      onChange={mockHandleChange}
    />
  );

  fireEvent.click(queryByText(/-/i));
  expect(mockHandleChange).toHaveBeenCalledTimes(1);
  expect(mockHandleChange).toHaveBeenCalledWith({
    target: {
      name: 'test',
      value: 1,
    },
  });

  fireEvent.click(queryByText(/\+/i));
  expect(mockHandleChange).toHaveBeenCalledTimes(2);
  expect(mockHandleChange).toHaveBeenCalledWith({
    target: {
      name: 'test',
      value: 3,
    },
  });
});

test('should CustomInputNumber disable plus feature', () => {
  const mockHandleChange = jest.fn();

  const { queryByText } = render(
    <CustomInputNumber
      min={0}
      max={4}
      step={1}
      name='test'
      value={4}
      disabled={false}
      onChange={mockHandleChange}
    />
  );

  fireEvent.click(queryByText(/\+/i));
  expect(mockHandleChange).toHaveBeenCalledTimes(0);
});

test('should CustomInputNumber disable minus feature', () => {
  const mockHandleChange = jest.fn();

  const { queryByText } = render(
    <CustomInputNumber
      min={0}
      max={4}
      step={1}
      name='test'
      value={0}
      disabled={false}
      onChange={mockHandleChange}
    />
  );

  fireEvent.click(queryByText(/-/i));
  expect(mockHandleChange).toHaveBeenCalledTimes(0);
});

test('should CustomInputNumber step props working', () => {
  const mockHandleChange = jest.fn();

  const { queryByText } = render(
    <CustomInputNumber
      min={0}
      max={4}
      step={2}
      name='test'
      value={4}
      disabled={false}
      onChange={mockHandleChange}
    />
  );

  fireEvent.click(queryByText(/-/i));
  expect(mockHandleChange).toHaveBeenCalledTimes(1);
  expect(mockHandleChange).toHaveBeenCalledWith({
    target: {
      name: 'test',
      value: 2,
    },
  });
});

test('should CustomInputNumber onBlur', () => {
  const mockOnblur = jest.fn();

  const { queryByText } = render(
    <CustomInputNumber
      min={0}
      max={4}
      step={2}
      name='test'
      value={2}
      disabled={false}
      onChange={() => {}}
      onBlur={mockOnblur}
    />
  );

  const valueNode = queryByText(/2/i);
  fireEvent.focus(valueNode);
  expect(mockOnblur).toHaveBeenCalledTimes(0);
  fireEvent.focusOut(valueNode);
  expect(mockOnblur).toHaveBeenCalledWith({
    target: {
      name: 'test',
      value: 2,
    },
  });
});
