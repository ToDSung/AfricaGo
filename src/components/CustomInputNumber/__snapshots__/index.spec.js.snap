// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should CustomInputNumber render active 1`] = `
<div
  className="flex gap-2 p-2"
>
  <div
    className="flex items-center justify-center rounded border text-base h-12 w-12  select-none text-4xl cursor-pointer  border-sky-600 text-sky-600"
    onClick={[Function]}
    onMouseDown={[Function]}
    role="minus"
  >
    -
  </div>
  <div
    className="flex items-center justify-center rounded border text-base 
              relative h-12 w-12 outline-none 
              border-sky-600 focus:border-2
            "
    onBlur={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    2
  </div>
  <div
    className="flex items-center justify-center rounded border text-base h-12 w-12  select-none  text-4xl   cursor-pointer border-sky-600  text-sky-600"
    onClick={[Function]}
    onMouseDown={[Function]}
    role="plus"
  >
    +
  </div>
</div>
`;

exports[`should CustomInputNumber render active 2`] = `
<div
  className="flex gap-2 p-2"
>
  <div
    className="flex items-center justify-center rounded border text-base h-12 w-12  select-none text-4xl cursor-pointer  border-sky-600 text-sky-600"
    onClick={[Function]}
    onMouseDown={[Function]}
    role="minus"
  >
    -
  </div>
  <div
    className="flex items-center justify-center rounded border text-base 
              relative h-12 w-12 outline-none 
              border-sky-600 focus:border-2
            "
    onBlur={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    3
  </div>
  <div
    className="flex items-center justify-center rounded border text-base h-12 w-12  select-none  text-4xl   cursor-pointer border-sky-600  text-sky-600"
    onClick={[Function]}
    onMouseDown={[Function]}
    role="plus"
  >
    +
  </div>
</div>
`;

exports[`should CustomInputNumber render disabled 1`] = `
<div
  className="flex gap-2 p-2"
>
  <div
    className="flex items-center justify-center rounded border text-base h-12 w-12  select-none text-4xl border-gray-600 text-gray-600"
    role="minus"
  >
    -
  </div>
  <div
    className="flex items-center justify-center rounded border text-base 
              relative h-12 w-12 outline-none 
              border-gray-600 text-gray-600
            "
  >
    2
  </div>
  <div
    className="flex items-center justify-center rounded border text-base h-12 w-12  select-none  text-4xl  border-gray-600 text-gray-600"
    role="plus"
  >
    +
  </div>
</div>
`;

exports[`should CustomInputNumber render disabled 2`] = `
<div
  className="flex gap-2 p-2"
>
  <div
    className="flex items-center justify-center rounded border text-base h-12 w-12  select-none text-4xl border-gray-600 text-gray-600"
    role="minus"
  >
    -
  </div>
  <div
    className="flex items-center justify-center rounded border text-base 
              relative h-12 w-12 outline-none 
              border-gray-600 text-gray-600
            "
  >
    3
  </div>
  <div
    className="flex items-center justify-center rounded border text-base h-12 w-12  select-none  text-4xl  border-gray-600 text-gray-600"
    role="plus"
  >
    +
  </div>
</div>
`;
